### PR TITLE
fix(client): use Proxy for "browser" import

### LINKED ISSUE #16153: Add documentation to "PrismaClient is unable to be run in the browser"
Many people, often not super experienced, encounter this. Although the error message is correct, we could probably do a better job at explaining what they are doing "wrong" and how they can solve it. The best solution would probably be to add this to the docs, and then add a link to the current error message (via pris.ly).

### PR DESCRIPTION
**Pull request by Jolg42**
[Internal Slack](https://prisma-company.slack.com/archives/C058VM009HT/p1691661424805119)

closes https://github.com/prisma/prisma/issues/16153

/integration

### CODE REVIEW

**Comment by Jolg42** [comment]
TODO update snapshots

File: helpers/compile/plugins/tscPlugin.ts [hunk 1]
```diff
@@ -70,7 +71,7 @@ export const tscPlugin: (emitTypes?: boolean) => esbuild.Plugin = (emitTypes?: b
 
     build.onStart(async () => {
       // we only call tsc if not in watch mode or in dev mode (they skip types)
-      if (options.incremental !== true && process.env.DEV !== 'true') {
+      if (process.env.WATCH !== 'true' && process.env.DEV !== 'true') {
```
Comment (millsp) [inline helpers/compile/plugins/tscPlugin.ts:74]: Unrelated fix.

File: packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap [hunk 2]
```diff
@@ -326,16 +326,21 @@ exports.Prisma.ModelName = {
...
+In case this error is unexpected for you, please report it via https://github.com/prisma/prisma/issues\`
+
   constructor() {
-    throw new Error(
-      \`PrismaClient is unable to be run \${runtimeDescription}.
-In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues\`,
-    )
+    return new Proxy(this, {
+      get(target, prop) {
+        throw new Error(this.message)
```
Comment (millsp) [inline packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap:338]: @Jolg42 initally had an extra case for $extends which I removed.

File: packages/client/src/generation/TSClient/common.ts [hunk 3]
```diff
@@ -83,24 +84,6 @@ Prisma.prismaVersion = {
...
-  // Deno
-  if ("Deno" in globalThis && typeof globalThis.Deno === "object") {
-    return "under the Deno runtime";
-  }
-  // Default to assuming browser
-  return "in the browser";
-})();
-`
-}
-
```
Comment (millsp) [inline packages/client/src/generation/TSClient/common.ts:103]: Old logic that I refactored so I could improve the error message.

File: packages/client/tests/e2e/_utils/standard.dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM node:16 as base
+FROM node:18 as base
```
Comment (millsp) [inline packages/client/tests/e2e/_utils/standard.dockerfile:1]: Needed for `fetch`

File: packages/client/src/generation/TSClient/common.ts (hunk 3 above)
Comment (millsp) [reply packages/client/src/generation/TSClient/common.ts:103]: We now use the `detect-runtime` package instead.

File: packages/client/tests/e2e/_utils/standard.dockerfile [hunk 5]
```diff
@@ -1,7 +1,9 @@
...
 
 FROM base as dependencies
 RUN npm i -g pnpm zx typescript jest ts-node esbuild esbuild-register @swc/jest
+RUN apt update
+RUN apt install iproute2 -y
```
Comment (aqrln) [inline packages/client/tests/e2e/_utils/standard.dockerfile:6]: is this needed for wrangler?

**Review by aqrln — APPROVED** [review]
Looks very good to me with one small question

File: packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap [hunk 6]
```diff
@@ -326,16 +312,35 @@ exports.Prisma.ModelName = {
...
-    throw new Error(
-      \`PrismaClient is unable to be run \${runtimeDescription}.
-In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues\`,
-    )
+    return new Proxy(this, {
+      get(target, prop) {
+        const runtime = detectRuntime()
+        const edgeRuntimeName = {
+          'workerd': 'Cloudflare Workers',
+          'deno': 'Deno and Deno Deploy',
```
Comment (janpio) [inline packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap:324]: Does this really apply to both Deno and Deno Deploy?

File: packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap [hunk 7]
```diff
@@ -326,16 +312,35 @@ exports.Prisma.ModelName = {
...
+          'netlify': 'Netlify Edge Functions',
+          'edge-light': 'Vercel Edge Functions',
```
Comment (janpio) [inline packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap:326]: Curious: Possible that this also would trigger for Vercel Edge Middlewares or different Next.js features users are not even aware that they are run as Vercel Edge Functions?

File: packages/client/tests/e2e/engine-not-found-error/binary-targets-incorrectly-pinned/tests/binary.ts [hunk 8]
```diff
@@ -16,14 +16,14 @@ Invalid \`prisma.user.create()\` invocation in
   4   const prisma = new PrismaClient()
   5 
 → 6   const result = prisma.user.create(
-Prisma Client could not locate the Query Engine for runtime "debian-openssl-1.1.x".
+Prisma Client could not locate the Query Engine for runtime "debian-openssl-3.0.x".
```
Comment (janpio) [inline packages/client/tests/e2e/engine-not-found-error/binary-targets-incorrectly-pinned/tests/binary.ts:19]: Caused by `node:18`?

File: packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap (hunk 6 above)
Comment (millsp) [reply packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/binary.test.ts.snap:324]: I don't think so, but there was code about it which had already been already accepted. I don't expect Deno to ever hit this path.

**Pull request merged**